src/tardev-snapshotter/src/snapshotter.rs

```diff
@@ -887,63 +887,110 @@ impl TarDevSnapshotter {
+                        retries,
+                        MAX_RETRIES,
+                        download_err
+                    );
+                    if retries >= MAX_RETRIES {
+                        return Err(Status::unknown(format!(
+                            "Failed to fetch layer image after {} attempts: {:?}",
+                            MAX_RETRIES, download_err
+                        )));
                     }
+
+                    warn!("Retrying layer image download...");
+                    continue; // Retry fetching the layer image
                 }
 
-                trace!("Appending index to {:?}", &base_name);
-                file.rewind().context("failed to rewind the file handle")?;
-                tarindex::append_index(&mut file).context("failed to append tar index")?;
+                // Process the layer
+                let process_result = tokio::task::spawn_blocking({
+                    let upstream_name = upstream_name.clone();
+                    let base_name = base_name.clone();
+                    let salt = salt.clone();
 
-                trace!("Appending dm-verity tree to {:?}", &base_name);
-                let root_hash = verity::append_tree::<Sha256>(&mut file, &salt)
-                    .context("failed to append verity tree")?;
+                    move || -> Result<_> {
+                        if layer_type == TAR_EXTENSION {
+                            info!("Renaming {:?} to {:?}", &upstream_name, &base_name);
+                            std::fs::rename(&upstream_name, &base_name)?;
+                        }
+                        let mut file = OpenOptions::new()
+                            .read(true)
+                            .write(true)
+                            .create(true)
+                            .truncate(layer_type == TAR_GZ_EXTENSION)
+                            .open(&base_name)?;
+                        if layer_type == TAR_GZ_EXTENSION {
+                            info!("Decompressing {:?} to {:?}", &upstream_name, &base_name);
+                            let compressed = fs::File::open(&upstream_name).map_err(|e| {
+                                let file_error = format!(
+                                    "Failed to open file {:?} for decompression: {:?}",
+                                    &upstream_name, e
+                                );
+                                error!("{}", file_error);
+                                anyhow::anyhow!(file_error)
+                            })?;
+
+                            let mut gz_decoder = flate2::read::GzDecoder::new(compressed);
+
+                            if let Err(e) = std::io::copy(&mut gz_decoder, &mut file) {
+                                let copy_error = format!("failed to copy payload from gz decoder {:?}", e);
+                                error!("{}", copy_error);
+                                return Err(anyhow::anyhow!(copy_error));
```

Comment:
This is the error we hit; should we trigger retry with a new download here as well? Or what are we doing to resolve it?

```
failed to extract image layer: failed to copy payload from gz decoder Error { kind: UnexpectedEof, message: \"failed to fill whole buffer\" }: unknown
```

Reply:
Yes, failing here will trigger a new download.